libraries/previousAuctionInfo/previousAuctionInfo.js

```diff
@@ -0,0 +1,130 @@
+import {on as onEvent} from '../../src/events.js';
+import { EVENTS } from '../../src/constants.js';
+
+export let previousAuctionInfoEnabled = false;
+let enabledBidders = [];
+export let winningBidsMap = {};
+
+export const resetPreviousAuctionInfo = () => {
+  previousAuctionInfoEnabled = false;
+  enabledBidders = [];
+  winningBidsMap = {};
+};
+
+export const enablePreviousAuctionInfo = (config, cb = initHandlers) => {
+  const { bidderCode, isBidRequestValid } = config;
+  const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidderCode);
+  if (!enabledBidder) enabledBidders.push({ bidderCode, isBidRequestValid, maxQueueLength: config.maxQueueLength || 10 });
+  if (previousAuctionInfoEnabled) return;
+  previousAuctionInfoEnabled = true;
+  cb();
+}
+
+export const initHandlers = () => {
+  onEvent(EVENTS.AUCTION_END, onAuctionEndHandler);
+  onEvent(EVENTS.BID_WON, onBidWonHandler);
+  onEvent(EVENTS.BID_REQUESTED, onBidRequestedHandler);
+};
+
+export const onAuctionEndHandler = (auctionDetails) => {
+  // eslint-disable-next-line no-console
+  console.log('onAuctionEndHandler', auctionDetails);
+
+  try {
+    let highestCpmBid = 0;
+    const receivedBidsMap = {};
+    const rejectedBidsMap = {};
+
+    if (auctionDetails.bidsReceived && auctionDetails.bidsReceived.length) {
+      highestCpmBid = auctionDetails.bidsReceived.reduce((highestBid, currentBid) => {
+        return currentBid.cpm > highestBid.cpm ? currentBid : highestBid;
+      }, auctionDetails.bidsReceived[0]);
+
+      auctionDetails.bidsReceived.forEach(bidReceived => {
+        receivedBidsMap[bidReceived.requestId] = bidReceived;
+      });
+    }
+
+    if (auctionDetails.bidsRejected && auctionDetails.bidsRejected.length) {
+      auctionDetails.bidsRejected.forEach(bidRejected => {
+        rejectedBidsMap[bidRejected.requestId] = bidRejected;
+      });
+    }
+
+    if (auctionDetails.bidderRequests && auctionDetails.bidderRequests.length) {
+      auctionDetails.bidderRequests.forEach(bidderRequest => {
+        const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidderRequest.bidderCode);
+
+        if (enabledBidder) {
+          bidderRequest.bids.forEach(bid => {
+            const isValidBid = enabledBidder.isBidRequestValid(bid);
+
+            if (!isValidBid) return;
+
+            const previousAuctionInfoPayload = {
+              bidderRequestId: bidderRequest.bidderRequestId,
+              minBidToWin: highestCpmBid?.cpm || 0,
+              rendered: 0,
+              transactionId: bid.ortb2Imp.ext.tid || bid.transactionId,
+              source: 'pbjs',
+              auctionId: auctionDetails.auctionId,
```

Comment:
Same issue as `transactionId` - this is hidden by default. This should either be omitted, or included only when [the `transmitTid` activity is allowed](https://github.com/prebid/Prebid.js/blob/a9958be99d7c9d431fe3d22929daf834009f3480/src/adapters/bidderFactory.js#L194)

Reply:
I think it should be debated whether any of this data is useful to bidders if there's no IDs. If I understand the purpose correctly, the receiver will join these logs with their bid logs so they can do stuff like "oh, when the browser is X and the country is Y, my bid has to be higher". In order to do that, I suspect that there's a minimal set of IDs or the join will be impossible.

Reply:
we can leave in `requestId` and `bidderRequestId`. They have the same domain (allow the same join) as `transactionId` / `auctionId`, the difference is they're not shared across bidders.

Or we can do what I think the first join would have to be - each request's prevAuctionInfo only pertains bids on the same ad unit.